zerver/openapi/zulip.yaml

```diff
@@ -6506,6 +6506,12 @@ paths:
         avoid generating very large HTTP responses. A maximum of 5000 messages
         can be obtained per request; attempting to exceed this will result in an
         error.
+
+        **Changes**: Added a new `message_ids` parameter, as an alternative method
+        of specifying which messages to fetch. `message_ids` is mutually
+        exclusive to `anchor`, `include_anchor`, `num_after`, `num_before` and
+        `use_first_unread_anchor` (deprecated). Note that this makes all the
+        parameters GET request to `/messages` API endpoint optional.
```

Comment:
Changes entries always need to mention the API feature level. I ended up reworking the docs significantly in this review pass; here's the diff:

``` diff
diff --git a/zerver/lib/narrow.py b/zerver/lib/narrow.py
index 01f35911c4..ad6df665c3 100644
--- a/zerver/lib/narrow.py
+++ b/zerver/lib/narrow.py
@@ -1442,8 +1442,12 @@ def fetch_messages(
 
     if client_requested_message_ids is not None:
         # We don't need to do any post-processing in this case.
+        if first_visible_message_id > 0:
+            visible_rows: Sequence[MessageRowT] = [r for r in rows if r[0] >= first_visible_message_id]
+        else:
+            visible_rows = rows
         return FetchedMessages(
-            rows=rows,
+            rows=visible_rows,
             found_anchor=False,
             found_newest=False,
             found_oldest=False,
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index 85286976bf..e78485ec37 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -6479,39 +6479,37 @@ paths:
       summary: Get messages
       tags: ["messages"]
       description: |
-        Fetch user's message history from a Zulip server.
+        This endpoint is the primary way to fetch a messages. It is used by all official
+        Zulip clients (e.g. the web, desktop, mobile, and terminal clients) as well as
+        many bots, API clients, backup scripts, etc.
 
-        This endpoint is the primary way to fetch a user's message history
-        from a Zulip server. It is useful both for Zulip clients (e.g. the
-        web, desktop, mobile, and terminal clients) as well as bots, API
-        clients, backup scripts, etc.
+        Most queries will specify a [narrow filter](/api/get-messages#parameter-narrow),
+        to fetch the messages matching any supported [search
+        query](/help/search-for-messages). If not specified, it will return messages
+        corresponding to the user's [combined feed](/help/combined-feed). There are two
+        ways to specify which messages matching the narrow filter to fetch:
+
+        - A range of messages, described by an `anchor` message ID (or a string-format
+          specification of how the server should computer an anchor to use) and a maximum
+          number of messages in each direction from that anchor.
+
+        - A rarely used variant (`message_ids`) where the client specifies the message IDs
+          to fetch.
+
+        The server returns the matching messages, sorted by message ID, as well as some
+        metadata that makes it easy for a client to determine whether there are more
+        messages matching the query that were not returned due to the `num_before` and
+        `num_after` limits.
 
         Note that a user's message history does not contain messages sent to
         channels before they [subscribe](/api/subscribe), and newly created
         bot users are not usually subscribed to any channels.
 
-        By specifying a [narrow filter](/api/get-messages#parameter-narrow),
-        you can use this endpoint to fetch the messages matching any search
-        query that is supported by Zulip's powerful full-text search backend.
-
-        In either case, you specify an `anchor` message (or ask the server to
-        calculate the first unread message for you and use that as the
-        anchor), as well as a number of messages before and after the anchor
-        message. The server returns those messages, sorted by message ID, as
-        well as some metadata that makes it easy for a client to determine
-        whether there are more messages matching the query that were not
-        returned due to the `num_before` and `num_after` limits.
-
-        We recommend setting `num_before` and `num_after` to no more than 1000, to
-        avoid generating very large HTTP responses. A maximum of 5000 messages
-        can be obtained per request; attempting to exceed this will result in an
-        error.
-
-        **Changes**: Added a new `message_ids` parameter, as an alternative method
-        of specifying which messages to fetch. `message_ids` is mutually
-        exclusive to `anchor`, `include_anchor`, `num_after`, `num_before` and
-        `use_first_unread_anchor` (deprecated). Note that this makes all the
-        parameters GET request to `/messages` API endpoint optional.
+        We recommend requesting at most 1000 messages in a batch, to avoid generating very
+        large HTTP responses. A maximum of 5000 messages can be obtained per request;
+        attempting to exceed this will result in an error.
+
+        **Changes**: The `message_ids` option is new in Zulip 10.0 (feature level 300).
       x-curl-examples-parameters:
         oneOf:
           - type: exclude
@@ -6660,13 +6658,16 @@ paths:
         - name: message_ids
           in: query
           description: |
-            A list of message IDs to fetch. Server will return messages matching
-            the narrow and the message IDs provided.
+            A list of message IDs to fetch. The server will return messages corresponding to the
+            subset of the requested message IDs that exist and the current user has access to,
+            potentially filtered by the narrow (if that parameter is provided).
+
+            It is an error to pass this parameter as well as any of the parameters involved in
+            specifying a range of messages: `anchor`, `include_anchor`, `use_first_unread_anchor`,
+            `num_before`, and `num_after`.
 
-            It is mutually exclusive with `anchor`, `include_anchor`, `num_before`,
-            and `num_after`, i.e., if `message_ids` is provided, then `anchor`,
-            `include_anchor`, `num_before` and `num_after` should not be provided
-            otherwise it will result in an error.
+            **Changes**: New in Zulip 10.0 (feature level 300). Previously, there was
+            no way to request a specific set of messages IDs.
           content:
             application/json:
               schema:
```